### PR TITLE
ready for review

### PR DESCRIPTION
**Pull request by skeydan**

### CODE REVIEW

File: _posts/2019-08-23-unet/unet.Rmd [hunk 1]
```diff
@@ -0,0 +1,437 @@
+---
+title: "Image segmentation with U-Net"
+description: > 
+ In image segmentation, every pixel of an image is assigned a class. Depending on the application, classes could be different cell types; or the task could be binary, as in "cancer cell yes or no?". Area of application nonwithstanding, the established neural network architecture of choice is U-Net. In this post, we show how to preprocess data and train a U-Net model on the Kaggle Carvana image segmentation data.
```
Comment (batpigandme) [inline _posts/2019-08-23-unet/unet.Rmd]:
```suggestion
 In image segmentation, every pixel of an image is assigned a class. Depending on the application, classes could be different cell types; or the task could be binary, as in "cancer cell yes or no?". Area of application notwithstanding, the established neural network architecture of choice is U-Net. In this post, we show how to preprocess data and train a U-Net model on the Kaggle Carvana image segmentation data.
```

File: _posts/2019-08-23-unet/unet.Rmd [hunk 2]
```diff
@@ -0,0 +1,437 @@
...
+  distill::distill_article:
+    self_contained: false
+preview: images/unet.png
+---
+
+```{r setup, include=FALSE}
+knitr::opts_chunk$set(echo = TRUE, eval = FALSE)
+```
+
+Sure, it is nice when I have a picture of some object, and a neural network can tell me what kind of object that is. More realistically, there might be several salient objects in that picture, and it tells me what they are, and where they are. The latter task especially (known as _object detection_) seems prototypical of contemporary AI applications that at the same time are fascinating intellectually and questionable ethically. It's different with the subject of this post: Successful _image segmentation_ has a lot of undeniably useful applications. For example, it is a sine qua non in medicine, neuroscience, biology and other life sciences.
```
Comment (batpigandme) [inline _posts/2019-08-23-unet/unet.Rmd]:
```suggestion
Sure, it is nice when I have a picture of some object, and a neural network can tell me what kind of object that is. More realistically, there might be several salient objects in that picture, and it tells me what they are, and where they are. The latter task (known as _object detection_) seems especially prototypical of contemporary AI applications that at the same time are intellectually fascinating and ethically questionable. It's different with the subject of this post: Successful _image segmentation_ has a lot of undeniably useful applications. For example, it is a sine qua non in medicine, neuroscience, biology and other life sciences.
```

Comment (batpigandme) [reply _posts/2019-08-23-unet/unet.Rmd]: Moved "especially" to be after the parenthetical (since you want the aside to be as close to what it's specifying as possible).

I also changed the order of "intellectually fascinating" (from fascinating intellectually) and "ethically questionable" (same flip-flop). I think it reads a bit more naturally that way, but, as always, feel free to override.

Comment (batpigandme) [reply _posts/2019-08-23-unet/unet.Rmd]: > It's different with the subject of this post:

I'm not 100% clear on what "it" refers to here.

File: _posts/2019-08-23-unet/unet.Rmd [hunk 3]
```diff
@@ -0,0 +1,437 @@
...
+
+So what, technically, is image segmentation, and how can we train a neural network to do it?
+
+## Image segmentation in a nutshell
+
+Say we have an image with a bunch of cats in it. In _classification_, the question is "what's that?", and the answer we want to hear is: "cat". In _object detection_, we again ask "what's that", but now that "what" is implicitly plural, and we expect an answer like "there's a cat, a cat and a cat, and they're here, here, and here" (imagine the network pointing, by means of drawing _bounding boxes_, i.e., rectangles around the detected objects). In _segmentation_, we want more: We want the whole image covered by "boxes" -- which aren't boxes anymore, but unions of pixel-size "boxlets" -- or put differently: __We want the network to label every single pixel in the image.__
```
Comment (batpigandme) [inline _posts/2019-08-23-unet/unet.Rmd]:
```suggestion
Say we have an image with a bunch of cats in it. In _classification_, the question is "what's that?", and the answer we want to hear is: "cat". In _object detection_, we again ask "what's that", but now that "what" is implicitly plural, and we expect an answer like "there's a cat, a cat, and a cat, and they're here, here, and here" (imagine the network pointing, by means of drawing _bounding boxes_, i.e., rectangles around the detected objects). In _segmentation_, we want more: We want the whole image covered by "boxes" -- which aren't boxes anymore, but unions of pixel-size "boxlets" -- or put differently: __We want the network to label every single pixel in the image.__
```

Comment (batpigandme) [reply _posts/2019-08-23-unet/unet.Rmd]: Added a comma so the "cat, cat, and a cat" has the same structure as "here, here, and here".

File: _posts/2019-08-23-unet/unet.Rmd [hunk 4]
```diff
@@ -0,0 +1,437 @@
...
+
+Here's an example from the paper we're going to talk about in a second. On the left is the input image (HeLa cells), next up is the ground truth, and third is the learned segmentation mask.
+
+```{r, eval=TRUE, echo=FALSE, layout="l-body-outset", fig.cap = "Example segmentation from Ronneberger et al. 2015."}
+knitr::include_graphics("images/mask.png")
+```
+
+Technically, a distinction is made between _class segmentation_ and _instance segmentation_. In class segmentation, referring to the "bunch of cats" example, there are two possible labels: Every pixel is either "cat" or "not cat". Instance segmentation is more difficult: Here every cat gets their own label. (As an aside, why should that be more difficult? Presupposing human-like cognition, it wouldn't be -- if I have the concept of a cat, instead of just "cattiness", I "see" there are two cats, not one. But depending on what a specific neural network relies on most -- texture, color, isolated parts -- those tasks may differ a lot in difficulty.)
```
Comment (batpigandme) [inline _posts/2019-08-23-unet/unet.Rmd]: I really like this aside, and (if it doesn't come up again later) you might consider making it _not_ an aside — give it a bit of room of its own.

File: _posts/2019-08-23-unet/unet.Rmd [hunk 5]
```diff
@@ -0,0 +1,437 @@
...
+
+The network architecture used in this post is adequate for _class segmentation_ tasks and should be applicable to a vast number of practical, scientific as well as non-scientific applications. Speaking of network architecture, how should it look?
+
+## Introducing U-Net
+
+Given their success in image classification, can't we just use a classic architecture like _Inception V[n]_, _ResNet_, _ResNext_ ... , whatever? The problem is, our task at hand -- labeling every pixel -- does not fit so well with the classic idea of a CNN. With convnets, the idea is to apply successive layers of convolution and pooling to build up feature maps of decreasing granularity, to finally arrive at an abstract level where we just say: "yep, a cat". The other side of the medail is, we lose detail information: To the final classification, it does not matter whether the five pixels in the top-left area are black or white.
```
Comment (batpigandme) [inline _posts/2019-08-23-unet/unet.Rmd]: I don't think you've spelled out CNN yet in this piece (maybe give it a hover-able "convolutional neural networks"?)

Comment (batpigandme) [reply _posts/2019-08-23-unet/unet.Rmd]: > medail

I'm not sure what this is. Are you saying something to the effect of _the downside_?

File: _posts/2019-08-23-unet/unet.Rmd [hunk 6]
```diff
@@ -0,0 +1,437 @@
...
+
+In practice, the classic architectures use (max) pooling or convolutions with `stride` > 1 to achieve those successive abstractions -- necessarily resulting in decreased spatial resolution.
+So how can we use a convnet and still preserve detail information? In their 2015 paper _U-Net: Convolutional Networks for Biomedical Image Segmentation_ [@RonnebergerFB15], Olaf Ronneberger et al. came up with what four years later, in 2019, is still the most popular approach. (Which is to say something, four years being a long time, in deep learning.)
```
Comment (batpigandme) [inline _posts/2019-08-23-unet/unet.Rmd]:
```suggestion
So how can we use a convnet and still preserve detail information? In their 2015 paper _U-Net: Convolutional Networks for Biomedical Image Segmentation_ [@RonnebergerFB15], Olaf Ronneberger et al. came up with what four years later, in 2019, is still the most popular approach. (Which is saying something, four years being a long time in deep learning.)
```

File: _posts/2019-08-23-unet/unet.Rmd [hunk 7]
```diff
@@ -0,0 +1,437 @@
...
+
+The idea is stunningly simple. While successive encoding (convolution / max pooling) steps, as usual, reduce resolution, the subsequent decoding -- we have to arrive at an output of  size same as the input, as we want to label every pixel! -- does not simply upsample from the most compressed layer. Instead, during upsampling, at every step we feed in information from the corresponding, in resolution, layer in the downsizing chain.
```
Comment (batpigandme) [inline _posts/2019-08-23-unet/unet.Rmd]:
```suggestion
The idea is stunningly simple. While successive encoding (convolution / max pooling) steps reduce resolution, the subsequent decoding -- we have to arrive at an output of  size same as the input, as we want to label every pixel! -- does not simply upsample from the most compressed layer. Instead, during upsampling, at every step we feed in information from the corresponding, in resolution, layer in the downsizing chain.
```

Comment (batpigandme) [reply _posts/2019-08-23-unet/unet.Rmd]: Took out that aside, since it's already a pretty complicated sentence.

File: _posts/2019-08-23-unet/unet.Rmd [hunk 8]
```diff
@@ -0,0 +1,437 @@
...
+
+For U-Net, really a picture says more than many words:
```
Comment (batpigandme) [inline _posts/2019-08-23-unet/unet.Rmd]:
```suggestion
For U-Net, a picture really says more than many words:
```

File: _posts/2019-08-23-unet/unet.Rmd [hunk 9]
```diff
@@ -0,0 +1,437 @@
...
+
+```{r, eval=TRUE, echo=FALSE, layout="l-body-outset", fig.cap = "U-Net architecture from Ronneberger et al. 2015."}
+knitr::include_graphics("images/unet.png")
+```
+
+
+Specifically, at each upsampling stage we _concatenate_ the output from the previous layer with that from its pendant in the compression stage. The final output is a _mask_ of size the original image, obtained via 1x1-convolution; no final dense layer is required, instead the output layer is just a convolutional layer with a single filter.
```
Comment (batpigandme) [inline _posts/2019-08-23-unet/unet.Rmd]:
```suggestion
At each upsampling stage we _concatenate_ the output from the previous layer with that from its pendant in the compression stage. The final output is a _mask_ of size the original image, obtained via 1x1-convolution; no final dense layer is required, instead the output layer is just a convolutional layer with a single filter. 
```

Comment (batpigandme) [reply _posts/2019-08-23-unet/unet.Rmd]: I also don't know what a pendant is, but I don't know very much about this topic.

File: _posts/2019-08-23-unet/unet.Rmd [hunk 10]
```diff
@@ -0,0 +1,437 @@
...
+# number of filters to start with, and number of classes to identify
+# see ?unet for more info
+model <- unet(input_shape = c(128, 128, 3))
+```
+
+So we have a model, and it looks like we'll be wanting to feed it 128x128 RGB images. Now how do we get these images?
+
+## The data
+
+To illustrate how applications arise even outside the area of medical research, we'll use as an example the Kaggle [Carvana Image Masking Challenge](https://www.kaggle.com/c/carvana-image-masking-challenge/overview). The task is to create a segmentation mask separating cars from background. For our current purpose, we only need `train.zip` and `train_mask.zip` from the [archive provided for download](https://www.kaggle.com/c/6927/download-all). In the following, we assume those have been extracted to a subdirectory called `data_raw`.
```
Comment (batpigandme) [inline _posts/2019-08-23-unet/unet.Rmd]: Have you considered making a repo with the data and code for the project? Was just thinking it could be nice to link to something that already has the directory structure, etc. you're describing.

File: _posts/2019-08-23-unet/unet.Rmd [hunk 11]
```diff
@@ -0,0 +1,437 @@
...
+```{r}
+example <- training_dataset %>% as_iterator() %>% iter_next()
+example$img %>% as.array() %>% as.raster() %>% plot()
+```
+
+```{r, eval=TRUE, echo=FALSE, layout="l-body", fig.cap = ""}
+knitr::include_graphics("images/transformed.png")
+```
+
+Now here's the complete preprocessing pipeline.
```
Comment (batpigandme) [inline _posts/2019-08-23-unet/unet.Rmd]:
```suggestion
Here's the complete preprocessing pipeline.
```

Comment (batpigandme) [reply _posts/2019-08-23-unet/unet.Rmd]: Since you start quite a few lines with "Now," I might recommend mixing it up by removing it in a few places.

File: _posts/2019-08-23-unet/unet.Rmd [hunk 12]
```diff
@@ -0,0 +1,437 @@
...
+______________________________________________________________________________________________
+conv2d_18 (Conv2D)             (None, 128, 128, 1) 65         conv2d_17[0][0]                 
+==============================================================================================
+Total params: 31,031,745
+Trainable params: 31,031,745
+Non-trainable params: 0
+______________________________________________________________________________________________
+```
+
+The "output shape" column shows the expected U-shape numerically: Width and height first go down, until we reach a minimum resolution of `8x8`; they then go up again, until we've reached the original resolution. At the same time, the number of filters first goes up, then goes down again, until in the output layer we have a single filter. You can also see the `concatenate` layers appending information that comes from "below" to information that comes "laterally".
```
Comment (batpigandme) [inline _posts/2019-08-23-unet/unet.Rmd]:
```suggestion
The "output shape" column shows the expected U-shape numerically: Width and height first go down, until we reach a minimum resolution of `8x8`; they then go up again, until we've reached the original resolution. At the same time, the number of filters first goes up, then goes down again, until in the output layer we have a single filter. You can also see the `concatenate` layers appending information that comes from "below" to information that comes "laterally."
```

File: _posts/2019-08-23-unet/unet.Rmd [hunk 13]
```diff
@@ -0,0 +1,437 @@
...
+
+What should be the loss function here? We're labeling each pixel, so each pixel contributes to the loss. We have a binary problem -- each pixel may be "car" or "background" -- so we want each output to be close to either 0 or 1. This makes _binary_crossentropy_ the adequate loss function.
```
Comment (batpigandme) [inline _posts/2019-08-23-unet/unet.Rmd]: Link for binary crossentropy maybe?

File: _posts/2019-08-23-unet/unet.Rmd (hunk 2 above)
Comment (skeydan) [reply _posts/2019-08-23-unet/unet.Rmd]: For the changes, sure, thanks! I had the feeling that some sentences were on the verge of being too complicated ... 
For this
> It's different with the subject of this post:
that was supposed to refer to the "questionnable ethically" thing ...  meaning, for segmentation, one tends to think more of medical applications (supposed to be "good"/useful) than of surveillance (as is the case with object detection) ... If you have an idea how to better convey this, I'd be happy to hear!

File: _posts/2019-08-23-unet/unet.Rmd (hunk 3 above)
Comment (skeydan) [reply _posts/2019-08-23-unet/unet.Rmd]: lol, sure, how could I ;-))

File: _posts/2019-08-23-unet/unet.Rmd (hunk 9 above)
Comment (skeydan) [reply _posts/2019-08-23-unet/unet.Rmd]: lol again! placeholder for sth to look up in the dictionary ... totally overlooked when proofreading ... wait ...
... that  will be "counterpart" :-)))

File: _posts/2019-08-23-unet/unet.Rmd (hunk 10 above)
Comment (skeydan) [reply _posts/2019-08-23-unet/unet.Rmd]: the data is 24 G :-(
but for the code, you're right, we have the same code in a vignette, I can add a footnote to say so, then people don't have to copy-paste

File: _posts/2019-08-23-unet/unet.Rmd (hunk 11 above)
Comment (skeydan) [reply _posts/2019-08-23-unet/unet.Rmd]: Totally. Actually I removed that and put it back again :-(
My normal procedure is "remove as many fill words as possible when proofreading".
Sometimes I'm wondering if it might sound too "hard", but if you say remove, it's remove :-)

File: _posts/2019-08-23-unet/unet.Rmd (hunk 13 above)
Comment (skeydan) [reply _posts/2019-08-23-unet/unet.Rmd]: will do!

File: _posts/2019-08-23-unet/unet.Rmd (hunk 5 above)
Comment (skeydan) [reply _posts/2019-08-23-unet/unet.Rmd]: oh yeah right ... actually was a typo, should have been "medal", but looks like it's unidiomatic anyway...

File: _posts/2019-08-23-unet/unet.Rmd (hunk 4 above)
Comment (skeydan) [reply _posts/2019-08-23-unet/unet.Rmd]: Thanks! Honestly I think that in this post, it's better to leave it  as an "aside".
But perhaps in the future, some time, you know, there could be posts directly aimed at discussing such things :-)
